Add slashing information to `ConsensusParams'

linked0 opened the pull request:
We add two slashing information to `ConsensusParams` such as `amount of penalty` and `slash threshold`. The `amount of penalty` is trivial. And the `slash threshold` is the criteria by which we punish the misbehaving validators. Currently, we punish the validators as soon as we detect them not to reveal preimages. But with this implementation, we can set more flexible policies about punishing the validators like slashing after two heights of missing pre-images or one-third of the cycle of missing pre-images.

Fixes #1486

linked0 committed:
Add consensus data for slashing protocol

By adding the data related to the slashing protocol to
the consensus parameter, we can manage the data more consistently.
We add the amount of penalty that the mis-behaving validators will lose
after being slashed.

AndrejMitrovic commented on source/agora/common/Config.d:
```diff
@@ -174,6 +175,9 @@ public struct NodeConfig
     public Duration block_catchup_interval = 20.seconds;
+
+    /// The amount of a penalty for slashed validators
+    public Amount slash_penalty_amount = Amount(100_000_000_000);
```
How about adding a new commit which implements a `coins` function which returns:

```D
Amount coins (in ulong num_coins)
{
    return Amount(UnitPerCoin.value * num_coins, true);
}
```

And then it can be used as:

```D
public Amount slash_penalty_amount = 10_000.coins;
```

MinFreezeAmount does something similar, but it uses the private `value` field. So I suggest adding a helper function. See here: https://github.com/bpfkorea/agora/blob/51466376cef32f26d04ab0d272d43951d4aec89e/source/agora/common/Amount.d#L41-L42

AndrejMitrovic replied:
Also it would be easier if the slashing value in the config is in coins rather than units.

linked0 replied:
> How about adding a new commit which implements a coins function which returns:
> Amount coins (in ulong num_coins)

Okay, it would be good. I will add it.

linked0 committed:
Add helper function for Amount

The function get the number of units and return an `Amount`
calculated with cents per unit and the number of coins.

linked0 commented:
Rebased and updated.
Ready for review.

AndrejMitrovic commented on source/agora/common/Config.d:
```diff
@@ -189,6 +189,9 @@ public struct NodeConfig
     /// How frequent the payments to Validators will be in blocks
     public uint payout_period = 144;
 
+    /// The threshold to slash the validators did not reveal pre-images
```
Could we document what threshold means? Something like "total number of blocks a node may refuse to reveal its preimage before it becomes slashed".

linked0 replied:
Updated.

AndrejMitrovic reviewed (approved):
Just a few comments.

AndrejMitrovic commented on source/agora/consensus/SlashPolicy.d:
```diff
@@ -105,7 +110,7 @@ public class SlashPolicy
         foreach (idx, utxo_key; keys)
         {
-            if (!this.hasRevealedPreimage(height, utxo_key))
+            if (!this.hasRevealedPreimage(this.heightForSlashing(height), utxo_key))
```
Am I understanding this correctly? This slashes the node if it doesn't reveal at least X preimages in advance, where X is the threshold?

linked0 replied:
The threshold, X is to specify how many blocks a validator can miss pre-images before being slashed. If it's set to 0 which is the default value, the validator is slashed as soon as the case of missing pre-images is detected. If 1 is set to the value, the validator can miss the pre-images for one block. This concept is based on the cases of missing preimages, which is described [the original design of the slashing protocol](https://github.com/bpfkorea/agora/issues/1076).

linked0 commented:
Would you please review this? It's only approved by @AndrejMitrovic.

hewison-chris reviewed (approved):
LGTM. The test failures are from PeriodicCatchup which I am currently fixing.

Geod24 commented on source/agora/common/Config.d:
```diff
@@ -197,6 +198,13 @@ public struct ConsensusConfig
     /// How frequent the payments to Validators will be in blocks
     public uint payout_period = 144;
+
+    /// Total number of blocks a node may refuse to reveal its preimage
+    /// before it becomes slashed
+    public uint slash_threshold = 0;
```
Is that ever going to be more than 0 ?

Geod24 commented on source/agora/common/Config.d:
```diff
@@ -474,6 +482,8 @@ private ConsensusConfig parseConsensusConfig (Node* node, in CommandLine cmdln)
     const validator_tx_fee_cut = get!(ubyte, "consensus", "validator_tx_fee_cut")(cmdln, node);
     const payout_period = get!(uint, "consensus", "payout_period")(cmdln, node);
+    auto slash_threshold = opt!(uint, "node", "slash_threshold")(cmdln, node);
+    auto slash_penalty_value = opt!(ulong, "node", "slash_penalty_amount")(cmdln, node);
```
You can use `get` here.

Geod24 commented on tests/system/node/0/config.yaml:
```diff
@@ -19,6 +19,8 @@ node:
 
 consensus:
   validator_cycle: 20
+  slash_threshold: 0
+  slash_penalty_amount: 10000
```
No need to write it if you're using the default value

linked0 replied:
Right. But It's just a template so that a user can change it.

Geod24 reviewed (commented):
What's the rationale for accepting missing pre-images ? I don't think we should accept any missing ones, because of how easy it is to set a watchtower, and because they can be revealed way ahead of time.

linked0 commented:
After discussing the necessity of  'slash_thresh`, we conclude that we don't need the feature. So I will remove the related code in this PR and request a review again.

linked0 commented:
Ready for review!

AndrejMitrovic commented:
LGTM.
